src/sagemaker_xgboost_container/algorithm_mode/dask_training.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import logging
+import os
+import socket
+from subprocess import Popen
+import sys
+import time
+
+import dask.dataframe as dask_dataframe
+from dask.dataframe import DataFrame
+from dask.dataframe import Series
+from dask.distributed import Client, wait
+from sagemaker_algorithm_toolkit import exceptions as exc
+import xgboost as xgb
+
+from sagemaker_xgboost_container.constants.xgb_constants import MODEL_NAME
+
+DASK_PATH = "/miniconda3/bin"
+SCHEDULER_PORT = "8786"
+
+WAIT_TIME_DAEMONS_STARTUP_SEC = 2
+WAIT_FOR_ALL_WORKERS_TIMEOUT_SEC = 20
+
+logger = logging.getLogger(__name__)
+
+
+def _start_daemons(scheduler_host_ip: str, current_host_ip: str):
+    cmd_start_scheduler = os.path.join(DASK_PATH, "dask-scheduler")
+    cmd_start_worker = os.path.join(DASK_PATH, "dask-cuda-worker")
+    schedule_conn_string = "tcp://{}:{}".format(scheduler_host_ip, SCHEDULER_PORT)
+
+    # We can set this based on number of vCPUs in the future.
+    # So far haven't seen improvement with > 4, during testing.
+    num_threads_per_worker_process = "4"
+    # Dask cuda worker API doc: https://docs.rapids.ai/api/dask-cuda/nightly/api.html
+    worker_cli_command = [cmd_start_worker, schedule_conn_string, "--no-dashboard",
+                          "--nthreads", num_threads_per_worker_process]
+    # Dask distributed scheduler API doc: https://docs.dask.org/en/stable/deploying-cli.html
+    scheduler_cli_command = [cmd_start_scheduler, "--no-dashboard"]
+    if scheduler_host_ip == current_host_ip:
+        Popen(scheduler_cli_command)
+    Popen(worker_cli_command)
+
+
+def _get_host_ip(host_name: str) -> str:
+    ip_wait_time = 5
+    counter = 0
+    host_ip = ""
+    from socket import gaierror
+    while counter < ip_wait_time and host_ip == "":
+        try:
+            host_ip = socket.gethostbyname(host_name)
+            break
+        except gaierror:
```

Comment:
Is this related to waiting for hosts to initialize or something else? I am wondering if this is really necessary since we do not really do this check for standard distributed jobs. We primarily just call `socket.gethostbyname` for each workers to check to see if each host is initialized.

Reply:
Ah ya, I see that in the tracker code. Will remove the explicit wait and just get the id.